lib/msal-node/src/retry/BaseRetryPolicy.ts

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import http from "http";
+import { Logger } from "@azure/msal-common";
+import { IHttpRetryPolicy } from "./IHttpRetryPolicy.js";
+
+/**
+ * Abstract base class for implementing HTTP retry policies.
+ * Classes that extend this base class must implement the `pauseForRetry` method.
+ */
+export abstract class BaseRetryPolicy implements IHttpRetryPolicy {
+    // Minimum backoff time in milliseconds. Can be overridden in derived classes.
+    protected minExponentialBackoff: number = 1000; // Default to 1 second
+    // Maximum backoff time in milliseconds. Can be overridden in derived classes.
+    protected maxExponentialBackoff: number = 4000; // Default to 4 seconds
+    // Maximum backoff time in milliseconds. Can be overridden in derived classes.
+    protected exponentialDeltaBackoff: number = 2000; // Default to 2 seconds
+
+    _isNewRequest: boolean;
+    set isNewRequest(value: boolean) {
+        this._isNewRequest = value;
+    }
+
+    abstract pauseForRetry(
+        httpStatusCode: number,
+        currentRetry: number,
+        logger: Logger,
+        retryAfterHeader?: http.IncomingHttpHeaders["retry-after"]
+    ): Promise<boolean>;
+
+    /**
+     * Calculates the number of milliseconds to sleep based on the `retry-after` HTTP header.
+     *
+     * @param retryHeader - The value of the `retry-after` HTTP header. This can be either a number of seconds
+     *                      or an HTTP date string.
+     * @returns The number of milliseconds to sleep before retrying the request. If the `retry-after` header is not
+     *          present or cannot be parsed, returns 0.
+     */
+    protected calculateLinearDelay(
```

Comment:
The base class contains logic related to its derived classes, which is a code smell.

Reply:
Consider using composition, not inheritance. In 99% of the cases, you can avoid class inheritance (i.e. avoid base classes, not interfaces)